<a href="https://colab.research.google.com/github/Joypamelakariuki/Joypamelakariuki/blob/main/Moringa_Data_Science_Prep_W4_Independent_Project_2022_02_Joy_Kariuki_DataReport.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Loading numpy and pandas to be used in the study

In [167]:
import numpy as np
import pandas as pd

In [168]:
#!curl http://bit.ly/autolib_dataset


In [169]:
#cars_dataset = 'https://archive.org/download/autolibdataset2/Autolib_dataset (2).csv'

## Loading the csv dataset to be analysed

In [170]:
cars_dataset = pd.read_csv('http://bit.ly/autolib_dataset')
#cars_dataset.head()

## Changing the dataset into a csv file

In [171]:
cars_data_csv = cars_dataset.to_csv('cars_data_csv',encoding='utf-8',index=False)
cars_data = pd.read_csv('cars_data_csv')
cars_data.head(10)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23
5                8 Boulevard Voltaire     0                0  ...    6     7       2
6                      37 rue Leblanc     0                0  ...    8    18      20
7         17 Rue des Luaps Prolongée     3                3  ...    4    22      13
8               34 avenue Jean Moulin     1                1  ...    2    22      58
9        41 boulevard de Rochechouart     6                6  ...    4    15       2

[10 rows x 25 columns]

-After loading the data, it is important to first understand how the data is. The colums names and the data, and how they can be used together.

In [172]:
cars_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   Displayed comment    111 non-null    object
 9   ID                   5000 non-null   object
 10  Kind                 5000 non-null   object
 11  Geo point            5000 non-null   object
 12  Postal code          5000 non-null   int64 
 13  Public name          5000 non-null   object
 14  Rental status        5000 non-null   object
 15  Scheduled at         47 non-null     object
 16  Slots 

## The  next step is to clean the data. This is done by dropping the columns that are not important or won't be  used. Among those that will be used include: Address, Bluecars, City, Month, Postal code, Station, Utilib, utilib 1.4, Hour.

In [173]:
cars_data_clean = cars_data.drop(['Cars','Charge Slots','Charging Status','Rental status','Slots','year','minute','Displayed comment','Geo point','Scheduled at'],axis=1)
cars_data_clean.head(5)

Address  Bluecar counter  ...  day  hour
0                 2 Avenue de Suffren                0  ...    8    11
1           145 Rue Raymond Losserand                6  ...    6     7
2    2 Avenue John Fitzgerald Kennedy                3  ...    3    20
3              51 Rue Eugène Oudiné                3  ...    4     4
4  6 avenue de la Porte de Champerret                3  ...    8    17

[5 rows x 15 columns]

The next step is to drop the duplicated rows for easier analysis

In [174]:
cars_data_clean = cars_data_clean.drop_duplicates()
cars_data_clean.head(5)

Address  Bluecar counter  ...  day  hour
0                 2 Avenue de Suffren                0  ...    8    11
1           145 Rue Raymond Losserand                6  ...    6     7
2    2 Avenue John Fitzgerald Kennedy                3  ...    3    20
3              51 Rue Eugène Oudiné                3  ...    4     4
4  6 avenue de la Porte de Champerret                3  ...    8    17

[5 rows x 15 columns]

## Analysis

Identify the most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.

In [175]:
##Changing the colums names for easier use
cars_data_clean.columns = ['Address', 'Bluecar_counter', 'Utilib_counter', 'Utilib1_counter', 'City', 'ID','Kind', 'Postal_code','public_name','station_type','Status','subscription_status','month','day','hour']
cars_data_clean.head(5)

Address  Bluecar_counter  ...  day  hour
0                 2 Avenue de Suffren                0  ...    8    11
1           145 Rue Raymond Losserand                6  ...    6     7
2    2 Avenue John Fitzgerald Kennedy                3  ...    3    20
3              51 Rue Eugène Oudiné                3  ...    4     4
4  6 avenue de la Porte de Champerret                3  ...    8    17

[5 rows x 15 columns]

In [176]:
#Picking the data for Paris city only
cars_data_Paris = cars_data_clean[cars_data_clean.City == 'Paris']
cars_data_Paris.head(20)

Address  Bluecar_counter  ...  day  hour
0                  2 Avenue de Suffren                0  ...    8    11
1            145 Rue Raymond Losserand                6  ...    6     7
3               51 Rue Eugène Oudiné                3  ...    4     4
4   6 avenue de la Porte de Champerret                3  ...    8    17
5                 8 Boulevard Voltaire                0  ...    6     7
6                       37 rue Leblanc                0  ...    8    18
8                34 avenue Jean Moulin                1  ...    2    22
9         41 boulevard de Rochechouart                6  ...    4    15
10                      14 rue Censier                0  ...    1     4
12  6 avenue de la Porte de Champerret                5  ...    5    12
13               34 avenue Jean Moulin                0  ...    8    21
14          8 Boulevard Richard Lenoir                0  ...    6     8
17              182 Boulevard Voltaire                6  ...    9    12
18               26 villa Croix-Nivert                5  ...    9    20
19                11 avenue René Coty                4  ...    9    21
20                2 Place de Catalogne                3  ...    5     6
24          15 avenue Émile Deschanel                2  ...    4    23
25              182 Boulevard Voltaire                0  ...    5    21
26    203 Rue du Faubourg-Saint-Martin                2  ...    3     9
28                        8 rue Danton                0  ...    2    23

[20 rows x 15 columns]

In [177]:
#Create another column for the usage of the blue car, if it was returned, its represented by a negative, if its picked, positive and 0 for none happening
cars_data_Paris['Bluecar_usage'] = cars_data_Paris.Bluecar_counter.diff()
cars_data_Paris

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Address  Bluecar_counter  ...  hour  Bluecar_usage
0                    2 Avenue de Suffren                0  ...    11            NaN
1              145 Rue Raymond Losserand                6  ...     7            6.0
3                 51 Rue Eugène Oudiné                3  ...     4           -3.0
4     6 avenue de la Porte de Champerret                3  ...    17            0.0
5                   8 Boulevard Voltaire                0  ...     7           -3.0
...                                  ...              ...  ...   ...            ...
4992                     3 avenue d'Ivry                0  ...    23            0.0
4994                    15 rue de Rocroy                0  ...    20            0.0
4996                   37 rue de Dantzig                4  ...    16            4.0
4997                      142 rue du Bac                1  ...     7           -3.0
4999                 5 Rue Henry Brisson                4  ...    10            3.0

[2795 rows x 16 columns]

# Picking the hour popular for picking up in Paris

In [178]:
#Picking the hour popular for picking up shared electricc car(Bluecar). We are picking a negative number since it represents cars being picked.
cars_data_Paris[(cars_data_Paris.City == 'Paris') & (cars_data_Paris.Bluecar_usage > 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(5)

hour
10    58
21    57
12    56
6     55
8     54
Name: hour, dtype: int64

From the data, the hour 10:00 is the most popular which is 10:00 am. for picking the cars.

# Most popular hour for returning the cars


In [179]:
#In  this case, we are not only considering Paris
cars_data_clean['Bluecar_usage'] = cars_data_clean.Bluecar_counter.diff()
cars_data_clean

Address  Bluecar_counter  ...  hour  Bluecar_usage
0                    2 Avenue de Suffren                0  ...    11            NaN
1              145 Rue Raymond Losserand                6  ...     7            6.0
2       2 Avenue John Fitzgerald Kennedy                3  ...    20           -3.0
3                 51 Rue Eugène Oudiné                3  ...     4            0.0
4     6 avenue de la Porte de Champerret                3  ...    17            0.0
...                                  ...              ...  ...   ...            ...
4995                 8 avenue Ménelotte                2  ...    11            2.0
4996                   37 rue de Dantzig                4  ...    16            2.0
4997                      142 rue du Bac                1  ...     7           -3.0
4998         2 avenue du Val de Fontenay                2  ...    17            1.0
4999                 5 Rue Henry Brisson                4  ...    10            2.0

[4968 rows x 16 columns]

In [180]:
cars_data_clean[(cars_data_clean.Bluecar_usage < 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(5)

hour
21    111
16    107
19    104
9      97
14     94
Name: hour, dtype: int64

21hours was the most popular hour for returning the cars which is 9:00pm.

# What station is the most popular overall





---



In [181]:
#Using the entire data not one with paris only
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')].groupby('public_name')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
public_name                     
Paris/Porte de Montrouge/8    13
Paris/Philippe Auguste/126    11
Paris/Courcelles/69           11
Sèvres/Wolfenbüttel/1       11
Paris/Mathis/35               11

Paris/Porte de Montrouge/8 is the most common station overall.

# What station is the most popular at the most picking hour?

In [182]:
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')& (cars_data_clean.Bluecar_usage > 0) ].groupby('public_name')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
public_name                                
Fontenay-aux-Roses/Robert Marchand/32    10
Paris/Porte de Montrouge/8                8
Fontenay-aux-Roses/Lombard/20             8
Paris/Mac Mahon/5                         7
Paris/Georges Mandel/42                   7

Fontenay-aux-Roses/Robert Marchand/32 is the popular station at the picking hour

# What postal code is the most popular for picking up Blue cars? 

In [183]:
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')& (cars_data_clean.Bluecar_usage > 0) ].groupby('Postal_code')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
Postal_code      
75015          96
75016          95
92000          80
75017          78
75020          76

75015 is the most popular postal code for picking up Blue cars.

# What postal code is the most popular for picking up Blue cars at the most popular picking hour? 

In [184]:
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')& (cars_data_clean.Bluecar_usage > 0) & (cars_data_clean.hour == 10)].groupby('Postal_code')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
Postal_code      
75020           6
75019           6
75016           4
75011           4
75017           4

75020 is the postal code.

The  most popular station do not belong to that postal code for both the overall and at the most popular picking hour.

# Checking the results again using Utilib data

##Checking the most popular hour of the day for picking Utilib in Paris

In [185]:
#Create another column for the usage of the blue car, if it was returned, its represented by a negative, if its picked, positive and 0 for none happening
cars_data_Paris['Utilib_counter_diff'] =cars_data_Paris.Utilib_counter.diff()
cars_data_Paris

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Address  ...  Utilib_counter_diff
0                    2 Avenue de Suffren  ...                  NaN
1              145 Rue Raymond Losserand  ...                  0.0
3                 51 Rue Eugène Oudiné  ...                  1.0
4     6 avenue de la Porte de Champerret  ...                 -1.0
5                   8 Boulevard Voltaire  ...                  0.0
...                                  ...  ...                  ...
4992                     3 avenue d'Ivry  ...                  0.0
4994                    15 rue de Rocroy  ...                  0.0
4996                   37 rue de Dantzig  ...                  0.0
4997                      142 rue du Bac  ...                  0.0
4999                 5 Rue Henry Brisson  ...                  0.0

[2795 rows x 17 columns]

In [186]:
#Picking the hour popular for picking up shared electricc car(utilib). We are picking a negative number since it represents cars being picked.
cars_data_Paris[(cars_data_Paris.City == 'Paris') & (cars_data_Paris.Utilib_counter_diff > 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(5)

hour
6     12
11    10
17     9
8      9
21     7
Name: hour, dtype: int64

The most popular time for picking the cars is 6:00 am

# What is the most popular hour for returning cars using the utilib data

In [187]:
#In  this case, we are not only considering Paris
cars_data_clean['Utilib_counter_diff'] = cars_data_clean.Utilib_counter.diff()
cars_data_clean

Address  ...  Utilib_counter_diff
0                    2 Avenue de Suffren  ...                  NaN
1              145 Rue Raymond Losserand  ...                  0.0
2       2 Avenue John Fitzgerald Kennedy  ...                  0.0
3                 51 Rue Eugène Oudiné  ...                  1.0
4     6 avenue de la Porte de Champerret  ...                 -1.0
...                                  ...  ...                  ...
4995                 8 avenue Ménelotte  ...                  0.0
4996                   37 rue de Dantzig  ...                  0.0
4997                      142 rue du Bac  ...                  0.0
4998         2 avenue du Val de Fontenay  ...                  0.0
4999                 5 Rue Henry Brisson  ...                  0.0

[4968 rows x 17 columns]

In [198]:
cars_data_clean[(cars_data_clean.City == 'Paris') & (cars_data_clean.Utilib_counter_diff < 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(5)

hour
19    12
23    10
22    10
14    10
6      9
Name: hour, dtype: int64

The most popular hour of returning is 19:00 which is 7:00pm

# The most popular station overall

In [189]:
#Using the entire data not one with paris only
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')].groupby('public_name')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
public_name                     
Paris/Porte de Montrouge/8    13
Paris/Philippe Auguste/126    11
Paris/Courcelles/69           11
Sèvres/Wolfenbüttel/1       11
Paris/Mathis/35               11

The station is the same as that of Bluecars

# What station is the most popular at the most picking hour?

In [190]:
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')& (cars_data_clean.Utilib_counter_diff > 0) ].groupby('public_name')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
public_name                        
Paris/Porte de Montrouge/8        4
Paris/Mac Mahon/5                 4
Meudon/Maréchal Juin/18          4
Paris/Mathis/35                   3
Gennevilliers/Petits marais/4     3

# What postal code is the most popular for picking up Utilib? 

In [191]:
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')& (cars_data_clean.Utilib_counter_diff > 0) ].groupby('Postal_code')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
Postal_code      
75016          18
75017          15
75015          14
75020          13
75014          11

The most common postal code is 75016

# What postal code is the most popular for picking up Utilib at the most popular picking hour? 

In [192]:
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')& (cars_data_clean.Utilib_counter_diff > 0) & (cars_data_clean.hour == 6)].groupby('Postal_code')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
Postal_code      
75020           3
75011           2
75015           2
75013           1
75014           1

The postal code is 75020 which is different from the Bluecars one.

# Checking the results using the Utilib 1.4 counter

##Checking the most popular hour of the day for picking Utilib1.4_counter in Paris

In [193]:
#Create another column for the usage of the blue car, if it was returned, its represented by a negative, if its picked, positive and 0 for none happening
cars_data_Paris['Utilib1_counter_diff'] =cars_data_Paris.Utilib1_counter.diff()
cars_data_Paris

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Address  ...  Utilib1_counter_diff
0                    2 Avenue de Suffren  ...                   NaN
1              145 Rue Raymond Losserand  ...                   0.0
3                 51 Rue Eugène Oudiné  ...                   0.0
4     6 avenue de la Porte de Champerret  ...                   0.0
5                   8 Boulevard Voltaire  ...                   0.0
...                                  ...  ...                   ...
4992                     3 avenue d'Ivry  ...                   0.0
4994                    15 rue de Rocroy  ...                   0.0
4996                   37 rue de Dantzig  ...                   0.0
4997                      142 rue du Bac  ...                   0.0
4999                 5 Rue Henry Brisson  ...                   0.0

[2795 rows x 18 columns]

In [195]:
cars_data_clean[(cars_data_clean.Utilib1_counter < 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(5)

Series([], Name: hour, dtype: int64)

# What is the most popular hour for returning cars using the utilib 1.4 data

In [196]:
#In  this case, we are not only considering Paris
cars_data_clean['Utilib1_counter_diff'] = cars_data_clean.Utilib1_counter.diff()
cars_data_clean

Address  ...  Utilib1_counter_diff
0                    2 Avenue de Suffren  ...                   NaN
1              145 Rue Raymond Losserand  ...                   0.0
2       2 Avenue John Fitzgerald Kennedy  ...                   2.0
3                 51 Rue Eugène Oudiné  ...                  -2.0
4     6 avenue de la Porte de Champerret  ...                   0.0
...                                  ...  ...                   ...
4995                 8 avenue Ménelotte  ...                   0.0
4996                   37 rue de Dantzig  ...                   0.0
4997                      142 rue du Bac  ...                   0.0
4998         2 avenue du Val de Fontenay  ...                   0.0
4999                 5 Rue Henry Brisson  ...                   0.0

[4968 rows x 18 columns]

In [197]:
cars_data_clean[(cars_data_clean.City == 'Paris') & (cars_data_clean.Utilib1_counter > 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head(5)

hour
21    18
9     17
5     16
10    16
22    15
Name: hour, dtype: int64

The most common returning hour is 9:00pm

# The most popular station overall

In [199]:
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')].groupby('public_name')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
public_name                     
Paris/Porte de Montrouge/8    13
Paris/Philippe Auguste/126    11
Paris/Courcelles/69           11
Sèvres/Wolfenbüttel/1       11
Paris/Mathis/35               11

# What postal code is the most popular for picking up Utilib 1.4 ? 

In [200]:
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')& (cars_data_clean.Utilib1_counter_diff > 0) ].groupby('Postal_code')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
Postal_code      
75017          29
75019          24
75015          22
75020          20
92000          19

The postal code is 75017

# What postal code is the most popular for picking up Utilib 1.4 at the most popular picking hour? 

In [201]:
cars_data_clean[(cars_data_clean.Kind == 'STATION') & (cars_data_clean.Status == 'ok')& (cars_data_clean.Utilib1_counter_diff > 0) & (cars_data_clean.hour == 6)].groupby('Postal_code')[['Kind']].count().sort_values(by = 'Kind', ascending  = False).head(5)


Kind
Postal_code      
75014           2
75015           2
75016           2
75020           2
92000           2

The postal code is 75014

In the analysis above, the most popular hour for picking the BlueCar in Paris is 10:00am while picking Utile is 6:00am. For Utilib1.4 it does not show the time. This can be as a result of inconsistent data.

The most popular hour for returning the cars is 9:00pm for Blue cars and Utilib1.4, 7:00 for Utilib.

The most popular station overall is common for the 2 cars.

The most popular station for the most picking hours is similar for Utilib and Utilin1.4 but different for Blue cars.
The postal code is different for the three cars.

The most popular postal code for the most picking hour is similar for both Redcars and Utile.
 The returning color is different for all the 3 car types.